### ISSUE TITLE
support bignum for number types not representable with float64

### ISSUE THREAD

**Issue by larkin** [nasa/openmct#1788]
Some customers have numbers that are larger than can be accurately represented in a float64.

Currently, if they transfer them over the wire as strings, they will work in all displays that don't parse the value-- e.g. fixed position displays, tables.

Even if the display does parse the value (e.g. a plot), it will work properly up until the point where the value exceeds what can be stored in float64.

We could add a format type (e.g. `bignum`) for large numbers and then see what it takes for displays that handle this type to represent this data.  Slightly uncertain how we could actually support this type in some display types.

**Comment by Costigan** [comment]
Pete,
  Does their problem with float64 stem from insufficient range or insufficient precision and range together?  If it's the latter, plots shouldn't be able to show the difference.  If it's the former, what in the world are they representing?  The number of atoms in the universe?  Oh, wait.  float64 is sufficient for that (to the precision we know).
  - Mark
________________________________

**Comment by larkin** [comment]
@Costigan high precision spacecraft clock (in nanoseconds) is the most legitimate case I've seen for this.

On the other hand, I've seen U64 used for counters that *probably* will never get that large; but since the data type is defined as a U64 they test their systems by generating data that matches the possible range of U64.

**Comment by akhenry** [comment]
Added help wanted label, not so much for implementation but for ideas. Others in the community may have faced the same problem.

**Comment by Kazmirchuk** [comment]
+1 on this
most modern spacecraft report telemetry with nanosecond precision, and AFAICS I have to round my timestamps to milliseconds to be able to display them in OpenMCT plots. Until a proper fix is available, I'd be grateful for hints on how I can workaround this issue.

**Comment by akhenry** [comment]
I think that there are two issues here. The JavaScript `Number` primitive type can represent 64 bit double precision values, so we're probably _not_ overflowing that right now (are we?). What we are definitely overflowing is the `Float32Array` type we are using specifically in our plots. I don't know how painful it would be to switch this to a `Float64Array` but it's something we will investigate.

The second issue is with representing time specifically. JavaScript cannot natively represent time with nanosecond precision. That said, I _think_ problems will only arise when going back and forth between human-readable and numerical representations of nanosecond precision date / timestamps, because we won't be able to use native functions, so there's a potential performance hit there. Our time API itself requires times to be converted to numbers though, and JavaScript Numbers give us plenty of range there, so the time API itself should work just fine with times in nanoseconds:
```
Number.MAX_VALUE
1.7976931348623157e+308
```

**Comment by mudinthewater** [comment]
Yeah this is a big pain in the ass for me on the serving side as well @akhenry. 


Looking at an example of a timestamp we'd have to be dealing with: 
'2015-148T15:22:41.34181226'


> data={'ts':['2015-148T15:22:41.34181226'],'value':[5]}
> df=pd.DataFrame(data)
> df
>                            ts  value
> 0  2015-148T15:22:41.34181226      5
> df['scet']=pd.to_datetime(df['ts'], cache=True, format='%Y-%jT%H:%M:%S.%f').view('int64')
> df
>                            ts  value                 scet
> 0  2015-148T15:22:41.34181226      5  1432826561341812260
> test=1432826561341812260/1000000
> test
> 1432826561341.8123

Python can't keep that many significant digits if I convert it to milliseconds, even though I _can_ convert it to an integer nanosecond timestamp. I can use something like decimal.Decimal to get the correct float value, but it's not really ideal either. So if we did want to support integer nanosecond timestamps I could do it, but if we support them via a float millisecond timestamp I can't. 

Javascript shares this imprecision for floating points: 

> var ts=Number(1432826561341812260)
> console.log(ts)
> console.log(ts/1000000)
> console.log(Number.MAX_SAFE_INTEGER)
> console.log(Number.MAX_VALUE)
Note how ts loses precision even before being divided: 
> 1432826561341812200
> 1432826561341.8123
> 9007199254740991
> 1.7976931348623157e+308

The maximum int64 is 9223372036854775807, which is Friday, April 11, 2262 11:47:16.854 PM. 
So unless there's way to eek out a few more digits of precision out of a javascript and python float, the only way I see this working is if the entire application is switched to nanosecond integer timestamps, which will max out in 2262 at which point it'll be someone else's problem I guess? But losing precision from Number(nanosecondtimestamp) would have to be solved too.

**Comment by mudinthewater** [comment]
@akhenry hate to bump such an old topic, but what would it take to switch to microseconds from milliseconds in MCT?

**Comment by WFsimon** [comment]
Also bumping: curious about greater than millisecond precision! Made a discussion post here https://github.com/nasa/openmct/discussions/8058

**Comment by akhenry** [comment]
@mudinthewater  Take at look at https://github.com/nasa/openmct/discussions/8058#discussioncomment-13211835. If this is a high priority for your users it's probably something we can support.
